digdag-standards/src/main/java/io/digdag/standards/operator/td/TdForEachOperatorFactory.java

```diff
@@ -130,20 +142,32 @@ protected String startJob(TaskExecutionContext ctx, TDOperator op, String domain
 
         private List<Config> fetchRows(TDJobOperator job)
         {
-            // TODO: handle netsplits
-
-            List<String> columnNames = job.getResultColumnNames();
-            return job.getResult(ite -> {
-                List<Config> rows = new ArrayList<>();
-                while (ite.hasNext()) {
-                    rows.add(row(columnNames, ite.next().asArrayValue()));
-                    if (rows.size() > Limits.maxWorkflowTasks()) {
-                        TaskLimitExceededException cause = new TaskLimitExceededException("Too many tasks. Limit: " + Limits.maxWorkflowTasks());
-                        throw new TaskExecutionException(cause, TaskExecutionException.buildExceptionErrorConfig(cause));
+            Config resultState = state.getNestedOrSetEmpty(RESULT);
+
+            try {
+                List<String> columnNames = job.getResultColumnNames();
+                return job.getResult(ite -> {
+                    List<Config> rows = new ArrayList<>();
+                    while (ite.hasNext()) {
+                        rows.add(row(columnNames, ite.next().asArrayValue()));
+                        if (rows.size() > Limits.maxWorkflowTasks()) {
+                            TaskLimitExceededException cause = new TaskLimitExceededException("Too many tasks. Limit: " + Limits.maxWorkflowTasks());
+                            throw new TaskExecutionException(cause, TaskExecutionException.buildExceptionErrorConfig(cause));
+                        }
                     }
+                    return rows;
+                });
+            }
+            catch (UncheckedIOException | TDClientException e) {
+                if (isDeterministicClientException(e)) {
+                    throw new TaskExecutionException(e, TaskExecutionException.buildExceptionErrorConfig(e));
                 }
-                return rows;
-            });
+                int retry = resultState.get(RETRY, int.class, 0);
+                resultState.set(RETRY, retry + 1);
```

Comment:
Yeah, ideally we'd have some structured primitives that we could use to abstract away the retry state-machine.